Factor out language_infot from languaget

peterschrammel opened the pull request:
Prerequisite for https://github.com/diffblue/cbmc/pull/2127 and https://github.com/diffblue/cbmc/pull/2098 to make access to language-specific methods more efficient.

Previously, using language-specific methods such as
from_expr() required creating a new languaget instance
which could be potentially costly.
This PR introduces a language_infot class that offers
language-specific methods that don't require parsing
functionality. language_infot is now used for language
registration (instead of language_entryt). The immutable
registered *_language_infot instances are shared among the
corresponding *_languaget instances that are requested by
the get_language_from_* functions.

tautschnig commented on src/clobber/CMakeLists.txt:
```diff
@@ -25,8 +25,6 @@ target_link_libraries(clobber-lib
 
 add_if_library(clobber-lib bv_refinement)
-add_if_library(clobber-lib specc)
-add_if_library(clobber-lib php)
```
Wishlist: Could those changes (and related ones) go in a separate commit, please?

tautschnig commented on src/clobber/clobber_parse_options.cpp:
```diff
@@ -38,8 +38,6 @@ Author: Daniel Kroening, kroening@kroening.com
 #include <cbmc/version.h>
 
-// #include "clobber_instrumenter.h"
-
```
Maybe a "Clobber: cleanup" commit?

peterschrammel replied:
Moved into a separate PR: https://github.com/diffblue/cbmc/pull/2159

tautschnig commented on src/goto-analyzer/CMakeLists.txt:
```diff
@@ -23,8 +23,6 @@ target_link_libraries(goto-analyzer-lib
 add_if_library(goto-analyzer-lib java_bytecode)
 add_if_library(goto-analyzer-lib jsil)
-add_if_library(goto-analyzer-lib specc)
-add_if_library(goto-analyzer-lib php)
```
To go in the cleanup commit.

tautschnig commented on src/clobber/Makefile:
```diff
@@ -31,16 +31,6 @@ CLEANFILES = clobber$(EXEEXT)
-  OBJ += ../php/php$(LIBEXT)
-  CP_CXXFLAGS += -DHAVE_PHP
-endif
-
```
Separate commit (see above).

tautschnig commented on src/ansi-c/ansi_c_language_info.h:
```diff
@@ -0,0 +1,45 @@
+  }
+  std::string description() const override
+  {
+    return "ANSI-C 99";
```
Correct copy&paste, but we have meanwhile gone well beyond "ANSI" C.

tautschnig commented on src/goto-cc/goto_cc_languages.cpp:
```diff
@@ -13,21 +13,13 @@ Author: CM Wintersteiger
-
-#ifdef HAVE_SPECC
-#include <specc/specc_language.h>
-#endif
```
The bottom part is for the cleanup commit.

tautschnig commented on src/goto-cc/goto_cc_languages.cpp:
```diff
@@ -13,21 +13,13 @@ Author: CM Wintersteiger
-
-  #ifdef HAVE_SPECC
-  register_language(new_specc_language);
-  #endif
```
The bottom part is for  the cleanup commit.

tautschnig commented on src/goto-diff/CMakeLists.txt:
```diff
@@ -25,7 +25,6 @@ target_link_libraries(goto-diff-lib
 
 add_if_library(goto-diff-lib java_bytecode)
 add_if_library(goto-diff-lib jsil)
-add_if_library(goto-diff-lib php)
```
Cleanup commit.

tautschnig commented on src/goto-diff/goto_diff_languages.cpp:
```diff
@@ -13,23 +13,14 @@ Author: Daniel Kroening, kroening@kroening.com
 
-#ifdef HAVE_SPECC
-#include <specc/specc_language.h>
-#endif
```
Cleanup commit.

tautschnig commented on src/goto-diff/goto_diff_languages.cpp:
```diff
@@ -13,23 +13,14 @@ Author: Daniel Kroening, kroening@kroening.com
-
-  #ifdef HAVE_SPECC
-  register_language(new_specc_language);
-  #endif
```
Cleanup commit.

tautschnig commented on src/jsil/jsil_language_info.h:
```diff
@@ -0,0 +1,42 @@
+
+  irep_idt id() const override
+  {
+    return ID_C;
```
Copy&paste error: "jsil" (should probably introduce an `ID_jsil`).

tautschnig commented on src/jsil/jsil_language_info.h:
```diff
@@ -0,0 +1,42 @@
+  }
+  std::string description() const override
+  {
+    return "ANSI-C 99";
```
Copy&paste error, was "Javascript Intermediate Language"

tautschnig commented on src/langapi/language_info.h:
```diff
@@ -0,0 +1,87 @@
+  virtual std::set<std::string> extensions() const
+  {
+    return std::set<std::string>();
+  }
```
Shouldn't these also be `= 0`? Those default implementations make no sense.

tautschnig commented on unit/analyses/dependence_graph.cpp:
```diff
@@ -8,15 +8,15 @@ Author: Chris Smowton, chris.smowton@diffblue.com
+#include <util/arith_tools.h>
+#include <util/c_types.h>
+#include <util/std_code.h>
+#include <util/symbol_table.h>
```
Nit pick: I'm not sure this is a value-adding change? We usually do grouping by directories.

tautschnig commented on unit/java_bytecode/java_object_factory/gen_nondet_string_init.cpp:
```diff
@@ -7,16 +7,16 @@
+#include <testing-utils/catch.hpp>
+#include <util/expr.h>
+#include <util/namespace.h>
+#include <util/std_code.h>
```
As above: does this add value?

tautschnig reviewed (changes requested):
I decided to ignore the don't-review-request... Just one genuine error spotted, otherwise nit-picks.

peterschrammel committed:
Factor out language_infot from languaget

Previously, using language-specific methods such as
from_expr() required creating a new languaget instance
which could be potentially costly.
This commit introduces a language_infot class that offers
language-specific methods that don't require parsing
functionality. language_infot is now used for language
registration (instead of language_entryt). The immutable
registered *_language_infot instances are shared among the
corresponding *_languaget instances that are requested by
the get_language_from_* functions.

peterschrammel commented:
@tautschnig, this is ready for re-review.

peterschrammel commented:
The TG bump needs to pass before merging this.

tautschnig commented on src/cbmc/cbmc_languages.cpp:
```diff
@@ -13,19 +13,19 @@ Author: Daniel Kroening, kroening@kroening.com
+#include <ansi-c/ansi_c_language_info.h>
+#include <cpp/cpp_language_info.h>
 
 #ifdef HAVE_JSIL
```
Would you be ok to add a commit to this PR that removes all occurrences of `HAVE_JSIL` from the codebase? It's not used consistently and thus doesn't really make sense. (Else it can happen in some later PR, once this is merged.)

tautschnig commented on src/langapi/language_info.h:
```diff
@@ -0,0 +1,81 @@
+    std::string &name,
+    const namespacet &ns) const;
+
+  virtual ~language_infot()
```
Nit pick: you might use `= default;`

peterschrammel committed:
Remove conditional inclusion of JSIL

peterschrammel commented:
@tautschnig, please check whether the last commit is what you intended.

tautschnig commented:
> @tautschnig, please check whether the last commit is what you intended.

Yes, thank you very much!

peterschrammel commented:
TG bump passing...

thk123 commented on src/ansi-c/ansi_c_language_info.cpp:
```diff
@@ -0,0 +1,52 @@
+
+Module:
+
+Author: Daniel Kroening, kroening@kroening.com
```
Should this be `Diffblue ltd`?

martin-cs replied:
If it is a new file written by a Diffblue employee, yes.

thk123 commented on src/cpp/cpp_language_info.h:
```diff
@@ -0,0 +1,48 @@
+  {
+  }
+
+  // conversion from expression into string
```
Either remove or turn into doxygen comment

thk123 commented on src/java_bytecode/java_bytecode_language_info.cpp:
```diff
@@ -0,0 +1,41 @@
+
+Module: Java Bytecode Language
+
+Author: Peter Schrammel
```
And this should probably also be Diffblue ltd?

thk123 commented on src/langapi/language_info.h:
```diff
@@ -0,0 +1,79 @@
+  /// \param ns: a namespace
+  /// \return false if conversion succeeds
+  virtual bool
+  from_expr(const exprt &expr, std::string &code, const namespacet &ns) const;
```
Why not `= 0`?

peterschrammel replied:
Yes, that could be considered.

thk123 commented on src/langapi/language_info.h:
```diff
@@ -0,0 +1,79 @@
+public:
+  // language id / description
+
+  virtual irep_idt id() const = 0;
```
Please document these

thk123 commented on src/langapi/mode.cpp:
```diff
@@ -17,45 +17,61 @@ Author: Daniel Kroening, kroening@cs.cmu.edu
-  for(const auto &language : languages)
-    if(mode == language.mode)
-      return language.factory();
+  for(const auto &language_info : languages)
```
Inner content spread over multiple lines so please wrap in braces

thk123 commented on src/langapi/mode.cpp:
```diff
@@ -17,45 +17,61 @@ Author: Daniel Kroening, kroening@cs.cmu.edu
+/// \return the language info (assumes that the language has been registered)
+const language_infot &get_language_info_from_mode(const irep_idt &mode)
+{
+  for(const auto &language_info : languages)
```
Inner content spread over multiple lines so please wrap in braces

thk123 replied:
Also mild preference to use above method

thk123 commented on src/langapi/mode.cpp:
```diff
@@ -103,30 +134,25 @@ std::unique_ptr<languaget> get_language_from_filename(
     std::string(filename, ext_pos+1, std::string::npos);
 
-  if(extension=="")
+  if(extension.empty())
```
Random fixes makes this long commit even harder to review - please avoid in future :slightly_smiling_face:

thk123 commented on src/langapi/mode.cpp:
```diff
@@ -103,30 +134,25 @@ std::unique_ptr<languaget> get_language_from_filename(
-        e_it++)
-      if(_stricmp(extension.c_str(), e_it->c_str())==0)
-        return l_it->factory();
+    for(const auto &ext : extensions)
```
to avoid having two variables called `extension` (one abbreviated) could you rename either the original or the new one to something more descriptive? (e.g. `language_extension` vs `file_extension` perhaps?)

thk123 reviewed (commented):
Just some minor changes, no need for re-review, will approve once I've checked TG bump

thk123 reviewed (approved):
TG pointer looks OK too

chrisr-diffblue reviewed (approved):
Looks like a reasonable cleanup to me.

martin-cs commented on src/cbmc/CMakeLists.txt:
```diff
@@ -16,6 +16,7 @@ target_link_libraries(cbmc-lib
     goto-instrument-lib
     goto-programs
     goto-symex
+    jsil
```
The "remove conditional inclusion of JSIL" patch seems to be unrelated to the PR.  It may well be a good idea (Do we need conditional compilation of features?  If so, what is the policy?) but  at the moment it is not explained or justified.

peterschrammel replied:
See thread above - requested by @tautschnig due to inconsistent use.

tautschnig replied:
I agree that it would better fit a separate PR, but that PR would necessarily depend on this one because of overlapping changes, hence adding it in here seemed ok to me.

martin-cs commented on src/langapi/mode.cpp:
```diff
@@ -17,45 +17,65 @@ Author: Daniel Kroening, kroening@cs.cmu.edu
+  }
+
+  INVARIANT(false, "unregistered language `" + id2string(mode) + "'");
+}
```
Unreachable?  Is this really an invariant?

peterschrammel replied:
You may call it that way, but I'd like to attach an error message to it ;)

martin-cs reviewed (approved):
Perhaps I am being unobservant but it seems like this could also tidy up some of the expr2c, expr2java, expr2internalstringinmadeuplanguage, expr2yetanotherformatbutwithdifferentprecedence mess.

mgudemann reviewed (approved):
LGTM

kroening commented:
langapi is dead and deprecated!

smowton reviewed (approved):
Looks great, thanks for making the effort!

smowton commented:
@kroening how would you prefer to do a language-specific thing, like "for the language relating to this symbol, get usual file extension" for example?

peterschrammel commented:
@smowton, the suggestion is to move the language-specific dispatch from the global language registry to the driver program. I.e. everything that requires something language-specific will be parametrised by a function that performs the language-specific dispatch. The presumption is that this will only be necessary in very few cases (show_goto_trace, etc). The advantage is that we can get rid of a global variable.

TGWDB commented:
This appears to be a nice to do, but is quite out of date. Closing since there seems to be a low likelihood of progress on this at the moment. Encourage the author to rebase and reopen if they would like this merged.

If you believe this closure is erroneous please reopen this PR.
